script.featherence.service/resources/lib/shared/commondownloader.py

```diff
@@ -0,0 +1,247 @@
+    except: resumable = False
+
+    if resumable:
+        pass
+
+    if content < 1:
+        if silent != True or exe != "": dialogok(title, file, localize(1446) + space + localize(21802), localize(13036, s=[localize(33003)]))
+        return "skip"
+
+    size = 1024 * 1024
+    mb   = content / (1024 * 1024)
+
+    if content < size:
+        size = content
+
+    total   = 0
+    notify  = 0
+    errors  = 0
+    count   = 0
+    resume  = 0
+    sleep   = 0
+
+    if silent == False or exe != "":
+        if dialogyesno(title + addonString_servicefeatherence(32138).encode('utf-8') + '[CR]' + file, addonString_servicefeatherence(32139).encode('utf-8') % (mb) + '[CR]' + addonString_servicefeatherence(32140).encode('utf-8'), nolabel=localize(222),  yeslabel=(12321)) == 1:
+			return "abort"
+
+    text = 'Download File Size : %dMB %s ' % (mb, dest)
+    printlog(title="doDownload", printpoint=printpoint, text=text, level=1, option="")
+
+    #f = open(dest, mode='wb')
+    f = xbmcvfs.File(dest, 'w')
+
+    chunk  = None
+    chunks = []
+
+    while True and not xbmc.abortRequested:
+        downloaded = total
+        for c in chunks:
+            downloaded += len(c)
+        percent = min(100 * downloaded / content, 100)
+        if percent >= notify:
+			if silent != True or exe != "":
+				xbmc.executebuiltin( "XBMC.Notification(%s,%s,%i)" % ( title + " - " + str(percent)+'%', dest, 10000))
+
+				text = 'Download percent : %s %s %dMB downloaded : %sMB File Size : %sMB' % (str(percent)+'%', dest, mb, downloaded / 1000000, content / 1000000)
+				printlog(title="doDownload", printpoint=printpoint, text=text, level=1, option="")
+
+				notify += percentinfo
+
+        chunk = None
+        error = False
+
+        try:        
+            chunk  = resp.read(size)
+            if not chunk:
+                if percent < 99:
+                    error = True
+                else:
+                    while len(chunks) > 0 and not xbmc.abortRequested:
+                        c = chunks.pop(0)
+                        f.write(c)
+                        del c
+
+                    f.close()
+                    #print '%s download complete' % (dest)
+                    if not xbmc.Player().isPlaying() and silent != True:
+                        xbmcgui.Dialog().ok(title, dest, '' , 'Download finished')
+                    return "ok"
+					
+        except Exception, e:
+            text = str(e)
+            printlog(title="doDownload", printpoint=printpoint, text=text, level=1, option="")
+			
+            error = True
+            sleep = 10
+            errno = 0
+
+            if hasattr(e, 'errno'):
+                errno = e.errno
+
+            if errno == 10035: # 'A non-blocking socket operation could not be completed immediately'
+                pass
+
+            if errno == 10054: #'An existing connection was forcibly closed by the remote host'
+                errors = 10 #force resume
+                sleep  = 30
```

Comment:
Somewhere else (there's a lot of files here) I'm pretty sure I saw a m3u8 download routine or a multi-part downloader with lots of sleeps in it. First I would question the need for a downloading a content file and then the issues with so many long sleeps.